src/daft-plan/src/ops/project.rs

```diff
@@ -22,20 +25,275 @@ impl Project {
         projection: Vec<Expr>,
         resource_request: ResourceRequest,
     ) -> Result<Self> {
-        let upstream_schema = input.schema();
+        // Factor the projection and see if there are any substitutions to factor out.
```

Comment:
What are the main reasons that we'd want to do this at construction time rather than optimization time?